### ISSUE TITLE
Upgraded to 3.4.7 deleted all bookmarks

### ISSUE THREAD

**Issue by bstin** [nextcloud/bookmarks#1305]
**Describe the bug**

**To Reproduce**
1. Deleted from 3.4.6 to 3.4.7 and now all bookmarks are gone. 
2. Tested on laptop + desktop each with Brave / Firefox and same results. 

**Expected behavior**
I can see the bookmarks in Nextcloud "Recent" page, but all folders are gone 

**Screenshots**

![2020-11-03_14-26_1](https://user-images.githubusercontent.com/10012474/98037240-e3672680-1de0-11eb-8ad4-d96da3b97adb.png)

**Desktop (please complete the following information):**
OS Linux

**Server (please complete the following information):**
 - OS: Debian 9
 - HTTP server: Apache 2
 - Database: MySQL 10.1
 - PHP version: v7.2
 - Nextcloud version: 19.0.4
 - Bookmarks app version: 3.4.7
 - Activated Nextcloud Apps: [the output of running `$ sudo -u www-data php occ app:list` in your command line]
 - Nextcloud configuration: [the output of running `sudo -u www-data php occ config:list system`]
 - Nextcloud external user backend: [e.g. ldap, or none]

**Additional context**

**Comment by stevesanda** [comment]
This happened to me as well. I can see my bookmarks listed in "Recent" but the folder structure I had in place is completely missing.

**Comment by potassium1338** [comment]
Can confirm this.  Is this just a displaying thing or are they bookmarks in fact missing? In the android app I still can see them...

OS: Ubuntu 20.04.1 LTS
PHP: 7.4.3
HTTP server: apache 2
Nextcloud: 19.0.3

**Comment by bstin** [comment]
Luckily I had another laptop that had not synced....I was able to put in airplane mode then export the bookmarks from Chrome. Then on I re-imported / sync into Nextcloud. Now all bookmarks are back...

I don't think the bookmarks are actually "deleted" per my original post...they in fact are accessible from "Recent" folder. I think somehow they loss their "Folder" or organizational links. Nonetheless, still a bit of a mess....

**Comment by pirovc** [comment]
My bookmarks are also gone. I can still see a list of all of them without any folder structure in the Untagged category. If I try to export them, I just get an empty .html.

**Comment by marcelklehr** [comment]
Sigh. I'm on it. The update broke the API. All data is still in the database.

**Comment by bstin** [comment]
Thank you Marcel!

**Comment by marcelklehr** [comment]
> The update broke the API. All data is still in the database.

Update: It turns out this is not true. The API works fine. What broke was a repair step that was supposed to remove orphaned folders, and instead removed all folders. :(

**Comment by bstin** [comment]
Well thanks for the quick update. I suppose restore from backup is only practical solution? 

If the folders were recreated manually how to get the “Recent Items” back into the newly created folders?

Maybe need to first run a script to assign all orphans to some default folder than can drag-drop from there? Or is it possible in UI to drag-drop from Recent items?

**Comment by NiteCrwlr** [comment]
Same happened to me. All Bookmarks gone. Tried to uninstall and install the Nextcloud-App again, but now I get an error:
"Client error: `GET https://github.com/nextcloud/bookmarks/releases/download/v4.0.4/bookmarks-4.0.4.tar.gz` resulted in a `404 Not Found` response: Not Found"
Then I found this thread here. Hmmm. I'm lucky that firefox creates backups of the bookmarks.

**Comment by daddydaycare** [comment]
Yup, me too bookmarks now gone from browser.  Cant see them in app, but it says i have a count total just none listed. Glad to see it is not just me. misery loves company.

**Comment by daddydaycare** [comment]
So are we just hosed?

**Comment by bstin** [comment]
One thing you should check....is I know Firefox, but I think Chrome/Brave also creates a Bookmarks.bak file somewhere in your profile directory....may be possible to simply rename that to Bookmarks and relaunch and have your Bookmarks back....

**Comment by xxKeith20xx** [comment]
> One thing you should check....is I know Firefox, but I think Chrome/Brave also creates a Bookmarks.bak file somewhere in your profile directory....may be possible to simply rename that to Bookmarks and relaunch and have your Bookmarks back....

Saved the day! I had to copy it over to another chromium browser to see it work for some reason, but I was then able to export html and then import into my browser with floccus.

**Comment by pirovc** [comment]
In Firefox I managed to restore my deleted bookmarks with: Ctrl+Shift+O -> Import and Backup -> Restore

**Comment by marcelklehr** [comment]
It turns out, only the root folders have been removed. I'm working on an update that will reinstate them along with their associated contents.

**Comment by marcelklehr** [comment]
> If the folders were recreated manually how to get the “Recent Items” back into the newly created folders?

Yeah, you should be able to move the existing bookmarks into the right folders.

**Comment by marcelklehr** [comment]
> So are we just hosed?

Short answer: Yes.

Long answer:

1. This was caused by a typo (:see_no_evil:).

2. I've recreated the damage -- there is no easy way to fix this with a script/repair step from my side.

3. v3.4.9 is out, which fixes the offending repair step (mainly to finally give access to a different fix that was screwed up by the typo. The new release doesn't fix the damage.)

4. I'm sorry.

**Comment by potassium1338** [comment]
Hi Marcel,
are you sure that it is now fixed?
I have installed 3.4.9. and restored a backup of the prefix_bookmarks_* tables. Although under the tags they are available the bookmarks in the root folder do not show up. Any ideas?

**Comment by bstin** [comment]
Hi @potassium1338  - I'm just a regular user, but I think trying to fix from the mysql side may be difficult....possibly broken / changed id's?

Its probably safer to fix from client side. Find a backup of bookmarks in Firefox / Chrome and restore that way via floccus

**Comment by marcelklehr** [comment]
> Any ideas?

My guess would be that the backup is not old enough

**Comment by potassium1338** [comment]
Is there a way to see in the admin section /log when a specific app has been updated?

**Comment by bstin** [comment]
Not sure, but on my install the dates on the folders under /var/www/nextcloud/apps matches update time.

**Comment by daddydaycare** [comment]
No worries Marcelklehr,
It was easy as, to disable bookmarks in NC,  Turn off sync in floccus, then restore a back up.  Thanks for your work. And more importantly, your honesty and approach to this. Kudos!

**Comment by ChrisWiegman** [comment]
Seeing this still in 4.0.5. I disabled the plugin and deleted its tables. Reactivated and synced and all worked for about an hour. Now all bookmarks are hidden in "recent" again. Using Floccus for sync.

**Comment by altepizza** [comment]
> I can still see a list of all of them without any folder structure in the Untagged category.

That's actually also working for me as well. Many of my tagged and untagged bookmarks are still there and even the API spits them out. Only the folders are missing and therefore they are not displayed in the nextcloud frontend... so I guess that I could write a short python script to move the orphans into a tmp folder

**Comment by altepizza** [comment]
Sorry for the double post, but I just created a script to get my orphaned bookmarks back into the GUI and wanted to share it with you: [link](https://github.com/altepizza/scripts)
USE AT YOUR OWN RISK

**Comment by daddydaycare** [comment]
Working good so far!
Thanks for updating app. 3.4.9

**Comment by reelsense** [comment]
Where is the official method to recover?

**Comment by marcelklehr** [comment]
@daddydaycare That usually indicates a server problem, try checking the nextcloud log

**Comment by marcelklehr** [comment]
@reelsense The official method is restoring a database backup.

**Comment by daddydaycare** [comment]
> @daddydaycare That usually indicates a server problem, try checking the nextcloud log

I opted to uninstall this and floccus. Will stick with Firefox sync for now.  unfortunately hit my time limit to troubleshoot this. Will check again in a few months as a cool idea to keep in NC.

**Comment by ChrisWiegman** [comment]
Do we have any updates here? I've had to go back to Firefox sync for now as this has been so thoroughly broken.

**Comment by marcelklehr** [comment]
I'm sorry for the long delay here, I had to take a break the last couple of weeks due to personal reasons, which I realize was unfortunate for everyone affected by this but not possible to avoid.

The fix for the bug that caused this has been released as part of v4.0.5 and 3.4.9 as far as I can tell.

I'm available to help troubleshoot any outstanding issues that resulted from this over at https://gitter.im/nextcloud-bookmarks/community

**Comment by marcelklehr** [comment]
Together with a user I worked out the following approach to getting your bookmarks app back on track, if you've been hit by this issue.

The plan is as follows:

1. get your current set of bookmarks out of the app
2. drop all bookmarks related database tables
3. recover all bookmarks related database tables from your backup
4. get the new bookmarks from the beginning back in

## 0.
Make a backup of your current nextcloud database.

## 1. get your current set of bookmarks out of the app
Run the following mysqldump command to create a backup of your current set of bookmarks (without folders).

```
mysqldump --no-create-info --insert-ignore --single-transaction db_name oc_bookmarks > newer_bookmarks.sql
```

## 2. drop all bookmarks related database tables
In your mysql shell run the following to remove all bookmarks related data from the datbase.
```
drop table oc_bookmarks; drop table oc_bookmarks_tags; drop table oc_bookmarks_folders; drop table oc_bookmarks_folders_public; drop table oc_bookmarks_root_folders; drop table oc_bookmarks_shared_to_shares; drop table oc_bookmarks_shares; drop table oc_bookmarks_shared_folders; drop table oc_bookmarks_tree;
```

## 3. recover all bookmarks related database tables from your backup
Assuming you have your nextcloud db backup in `nextcloud-db-backup.sql`, we are extracting the bookmarks related tables from that file into a new file:

```
sed -n -e '/CREATE TABLE.*`oc_bookmarks.*`/,/Table structure for table/p' nextcloud-db-backup.sql > old_bookmarks_tables.sql
```

Then we're copying the comment headers from the backup file over to the start of  `olc_bookmarks_tables.sql` to make it work.

Now, we're importing those tables into the database:

```
mysql nextcloud_db_name < old_bookmarks_tables.sql
```

(`nextcloud_db_name` is the name of your database.)

## 4. get the new bookmarks from the beginning back in
This will re-insert any newer bookmarks that did not yet exist when the backup was done.

```
mysql nextcloud_db_name < newer_bookmarks.sql
```

## Done.
Now you should be able to see your folder structure in the bookmarks app again. Any bookmarks created after the folders were destroyed should be available in 'Recent bookmarks'. You can now move them to a folder of your choosing.

I'm closing this for now. If you have troubles with this or anything else, feel free to comment here, open a new issue, or contact me at https://gitter.im/nextcloud-bookmarks/community or over at https://cloud.nextcloud.com/call/u52jcby9
